File: Minimum sum partition - GFG/minimum-sum-partition.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;
class GfG
{
    public static void main(String args[])
        {
            Scanner sc = new Scanner(System.in);
            int t = sc.nextInt();
            while(t-->0)
                {
                    int n = sc.nextInt();
                    int A[] = new int[n];
                    for(int i = 0;i<n;i++)
                        A[i] = sc.nextInt();
                    Solution ob = new Solution();
                    System.out.println(ob.minDifference(A,n));
                }
        }
}    
// } Driver Code Ends


//User function Template for Java

class Solution
{
    private int util(int arr[], int totSum, int currSum,  int currInd,int[][] dp){
        int n = arr.length;
        if(currInd==n) return Math.abs(2*currSum-totSum);
        if(dp[currInd][currSum]!=-1){
            return dp[currInd][currSum];
        }
        return dp[currInd][currSum]=Math.min(util(arr,totSum,currSum+arr[currInd],currInd+1,dp), util(arr,totSum,currSum,currInd+1,dp));
    }
	public int minDifference(int arr[], int n) 
	{ 
	    int sum=0;
	    for(int a:arr){
	        sum+=a;
	    }
	    int[][] dp=new int[n][sum+1];
	    for(int i=0;i<n;i++){
	        Arrays.fill(dp[i],-1);
	    }
	    return util(arr,sum,0,0,dp);
	} 
}

File: Minimum sum partition - GFG/minimum-sum-partition.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution{

  public:
	int minDifference(int arr[], int n)  { 
	    int sum = 0;
	    for(int i=0;i<n;i++) sum+=arr[i];
	    
	    vector<vector<int>> dp(n+1,vector<int>(sum+1,0));
	    dp[0][0] = 1;
	    for(int i=1;i<=n;i++){
	        for(int j=1;j<=sum;j++){
	            if(arr[i-1]<=j) dp[i][j] = dp[i-1][j-arr[i-1]];
	            dp[i][j] = dp[i-1][j] || dp[i][j];
	        }
	    }
	    int ans = sum;
	    for(int i=0;i<=sum;i++){
	        if(dp[n][i]){
	            ans = min(ans, abs(2*i-sum));
	            
	        }
	        
	    }
	    return ans;
	} 
};


//{ Driver Code Starts.
int main() 
{
   
   
   	int t;
    cin >> t;
    while (t--)
    {
        int n;
        cin >> n;

        int a[n];
        for(int i = 0; i < n; i++)
        	cin >> a[i];

       

	    Solution ob;
	    cout << ob.minDifference(a, n) << "\n";
	     
    }
    return 0;
}
// } Driver Code Ends